### PR TITLE
Fix S1168 FP: Does not respect nullable annotations

### LINKED ISSUE #6825: Expand the ShimLayer with ITypeSymbol.NullableAnnotation
[ITypeSymbol.NullableAnnotation](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.itypesymbol.nullableannotation?view=roslyn-dotnet-4.3.0) is currently not part of the ShimLayer. Adding it, will make checking for Nullables easier and more robust.
See also:
https://github.com/SonarSource/sonar-dotnet/issues/3273#issuecomment-606693989
https://github.com/SonarSource/sonar-dotnet/pull/6753#discussion_r1107184744

### LINKED ISSUE #6878: Fix S1168 FP: Does not respect nullable annotations
### Description

The rule S1168 raises false positive when then nullable annotation is enabled.

### Repro steps

```
public interface IInterface
{
    public byte[]? GetNullableFileContent(string itemId);
    public byte[] GetFileContent(string itemId);
}

public class Implementation : IInterface
{
    public byte[]? GetNullableFileContent(string itemId) => null; // Should not raise because nullable is allowed
    public byte[] GetFileContent(string itemId) => null; // Should raise because null is unexpected
}
```

### Expected behavior

Should not raise when nullable is enabled and the syntax is used.
For me there is a big difference between an empty list and a not-existing list. Those are two completely different statements.
On the one hand, not existing means that there is no information about it. On the other hand, an empty list means that, for example, no information has been entered.

For example:
Returning null means "there is no file". Returning an empty array means "there is a file but this file is empty".

I mean rule is very useful when it is required that the value is notnull. When the annotation allows nullable values it is very annoying because the nullables allows the difference between "there is nothing" and "there is something empty".

I think this rule should not be triggered when nullable is enabled and used because the compiler (on newer versions) warns when not checking a nullable value

### Actual behavior

Rule will be raised when nullable is enabled and used

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #6878
Fixes #6825
Add a check on the return type to check nullability.

### CODE REVIEW

**Comment by Corniel** [comment]
I know the current solution is not something that should be shipped. In `SonarAnalyzer.CFG` the following code exists:

``` C#
// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

namespace StyleCop.Analyzers.Lightup
{
    public readonly record struct NullabilityInfo
    {
       // ..
    }
}
```
Obviously that should become available in `SonarAnalyzer.Common` too. The question is how? Furthermore I noticed that in common there is the Helpers namespace with `TypeHelper`, and there is the Extensions namespace with `NamedTypeInfoExtensions`. Should that be updated?

**Comment by martin-strecker-sonarsource** [comment]
> Obviously that should become available in SonarAnalyzer.Common too.

We track the implementation for this in https://github.com/SonarSource/sonar-dotnet/issues/6825 The idea is to do something 

> Should that be updated?

Yes. Something like in TypeInfoExtensions needs to be created:

```cs
public static class ITypeSymbolExtensions
{
    public NullableAnnotation NullableAnnotation(this ITypeSymbol symbol) { }
}
```

It is a bit easier as `TypeInfoExtensions.Nullability(this TypeInfo typeInfo)` but still needs the enum to enum conversion. Maybe one of the `LightupHelpers.CreateSyntaxPropertyAccessor` already does what we need (there is an overload that takes a type and calls Convert).

**Comment by Corniel** [comment]
> It is a bit easier as `TypeInfoExtensions.Nullability(this TypeInfo typeInfo)` but still needs the enum to enum conversion. Maybe one of the `LightupHelpers.CreateSyntaxPropertyAccessor` already does what we need (there is an overload that takes a type and calls Convert).

I could not find anything outside the CFG project. (I searched for `NullableAnnotation`). Do you know how in the past the LightupHelpers have been extended?

**Comment by martin-strecker-sonarsource** [comment]
You can have a look at this PR #6099. It shouldn't be as complicated in this case. A simple call to `LightupHelpers.CreateSyntaxPropertyAccessor` should return a proper `Func` if I'm not mistaken.

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/LightupHelpers.cs [hunk 1]
```diff
@@ -1,686 +1,682 @@
...
-                throw new InvalidOperationException();
-            }
+        Expression body = Expression.Call(instance, property.GetMethod);
 
-            var property = type.GetTypeInfo().GetDeclaredProperty(propertyName);
-            if (property == null)
-            {
-                return FallbackAccessor;
-            }
+        if (!typeof(TProperty).GetTypeInfo().IsAssignableFrom(property.PropertyType.GetTypeInfo()))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ShimLayer/LightupHelpers.cs:319]: 👍That is exactly what we need here!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs [hunk 2]
```diff
@@ -18,68 +18,70 @@
...
+        && !method.ReturnType.DerivesFrom(KnownType.System_Xml_XmlNode)
+        && method.ReturnType.DerivesOrImplementsAny(CollectionTypes)
+        && method.ReturnType.NullableAnnotation() != NullableAnnotation.Annotated;
 
-            return methodSymbol != null
-                && !methodSymbol.ReturnType.Is(KnownType.System_String)
-                && !methodSymbol.ReturnType.DerivesFrom(KnownType.System_Xml_XmlNode)
-                && methodSymbol.ReturnType.DerivesOrImplementsAny(CollectionTypes);
-        }
+    private static IMethodSymbol GetReturnMethod(SonarSyntaxNodeReportingContext context)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs:78]: This method can be inlined.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs [hunk 3]
```diff
@@ -18,68 +18,70 @@
...
             }
         }
+    }
 
-        private static bool IsReturningCollection(SonarSyntaxNodeReportingContext context)
-        {
-            var symbol = context.SemanticModel.GetDeclaredSymbol(context.Node);
-            var methodSymbol = (symbol as IPropertySymbol)?.GetMethod ?? symbol as IMethodSymbol;
+    private static bool IsReturningCollection(SonarSyntaxNodeReportingContext context) =>
+        GetReturnMethod(context) is { } method
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs:72]: `ReturnType` can be extracted directly
```suggestion
        GetReturnMethod(context) is { ReturnType: { } returnType }
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Nice work, @Corniel! I just have some minor suggestions.

File: analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs [hunk 4]
```diff
@@ -18,176 +18,179 @@
...
 
-namespace SonarAnalyzer.Helpers
+namespace SonarAnalyzer.Helpers;
+
+internal static class TypeHelper
 {
-    internal static class TypeHelper
-    {
-        #region TypeKind
+    private static readonly Func<ITypeSymbol, NullableAnnotation> NullableAnnotationAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<ITypeSymbol, NullableAnnotation>(typeof(ITypeSymbol), nameof(NullableAnnotation));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs:25]: This line is too long.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs (hunk 2 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs]: I'm not sure if you thought about the same change, but I've changed the method to bet `GetType()`, and I think it makes matters even easier.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. @Corniel thank you for your contribution!

**Comment by Corniel** [comment]
@martin-strecker-sonarsource Due to the updated indenting, some existing lines popped up that are not covered by existing unit tests. What to do?
1. Remove lines
2. Add tests
3. Ignore coverage

**Comment by martin-strecker-sonarsource** [comment]
3. Ignore coverage

The coverage is only about the non-whitespace changes and this case: `_ => null,` We don't care about covering this case as it is just defensive coding.

**Comment by Corniel** [comment]
> 3. Ignore coverage
> 
> 
> The coverage is only about the non-whitespace changes and this case: `_ => null,` We don't care about covering this case as it is just defensive coding.

I fully agree, but it is not my call, therefor the question.

**Comment by martin-strecker-sonarsource** [comment]
Peach validation:
* aspnet-core: ~30 FPs fixed 
* dnspy: ~40FPs fixed
* dnspy:extensions: ~120 FPs fixed
* dotnet-runtim: ~140 FPs fixed
* elasticsearch: 5 FPs fixed
* embedio: 3 FPs fixed
* hot-chocolate: 18 FPs fixed
* nodatime: 2 FPs fixed
* nunit: 4 FPs fixed

That's an impressive list @Corniel Thank you for the contribution!

**Pull request merged**